pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -2164,6 +2168,190 @@ public boolean instanceExists(String instanceName) {
+    if (segmentsTo == null || segmentsTo.isEmpty()) {
+      String errorMsg = String
+          .format("'segmentsTo' cannot be null or empty (tableName = '%s', segmentsFrom = '%s', segmentsTo = '%s'",
+              tableNameWithType, segmentsFrom, segmentsTo);
+      throw new IllegalArgumentException(errorMsg);
+    }
+
+    // segmentsFrom can be empty in case of the initial upload
+    if (segmentsFrom == null) {
+      segmentsFrom = new ArrayList<>();
+    }
+
+    // Check that all the segments from 'segmentsFrom' exist in the table
+    Set<String> segmentsForTable = new HashSet<>(getSegmentsFor(tableNameWithType));
+    if (!segmentsForTable.containsAll(segmentsFrom)) {
+      String errorMsg = String.format(
+          "Not all segments from 'segmentsFrom' are available in the table. (tableName = '%s', segmentsFrom = '%s', "
+              + "segmentsTo = '%s', segmentsFromTable = '%s')", tableNameWithType, segmentsFrom, segmentsTo,
+          segmentsForTable);
+      throw new IllegalArgumentException(errorMsg);
+    }
+
+    try {
+      final List<String> finalSegmentsFrom = segmentsFrom;
+      DEFAULT_RETRY_POLICY.attempt(() -> {
+        // Fetch the segment lineage metadata
+        ZNRecord segmentLineageZNRecord =
+            SegmentLineageAccessHelper.getSegmentLineageZNRecord(_propertyStore, tableNameWithType);
+        SegmentLineage segmentLineage;
+        int expectedVersion = -1;
+        if (segmentLineageZNRecord == null) {
+          segmentLineage = new SegmentLineage(tableNameWithType);
+        } else {
+          segmentLineage = SegmentLineage.fromZNRecord(segmentLineageZNRecord);
+          expectedVersion = segmentLineageZNRecord.getVersion();
+        }
+
+        // Check that the batchId doesn't exists in the segment lineage
+        if (segmentLineage.getLineageEntry(batchId) != null) {
+          String errorMsg = String.format("BatchId (%s) already exists in the segment lineage.", batchId);
+          throw new IllegalArgumentException(errorMsg);
+        }
+
+        for (String lineageEntryId : segmentLineage.getLineageEntryIds()) {
+          LineageEntry lineageEntry = segmentLineage.getLineageEntry(lineageEntryId);
+
+          // Check that any segment from 'segmentsFrom' does not appear on the left side.
+          if (lineageEntry.getSegmentsFrom().stream().anyMatch(finalSegmentsFrom::contains)) {
+            String errorMsg = String.format(
+                "It is not allowed to merge segments that are already merged. (tableName = %s, segmentsFrom from "
+                    + "existing lineage entry = %s, requested segmentsFrom = %s)", tableNameWithType,
+                lineageEntry.getSegmentsFrom(), finalSegmentsFrom);
+            throw new IllegalArgumentException(errorMsg);
+          }
+
+          // Check that merged segments name cannot be the same for different lineage entry
+          if (lineageEntry.getSegmentsTo().stream().anyMatch(segmentsTo::contains)) {
```

Comment:
Need to check both `Collections.disjoint(lineageEntry.getSegmentsFrom(), segmentsTo))` and `Collections.disjoint(lineageEntry.getSegmentsTo(), segmentsTo))`

Reply:
yes, we need to check both.